### PR TITLE
WIP: COMMON: Added tracy profiler support

### PR DESCRIPTION
**Pull request by mgerhardy**
Frame based, on-demand profiler: https://github.com/wolfpld/tracy

https://github.com/wolfpld/tracy/releases/download/v0.7.4/tracy.pdf

# Open Tasks

* [ ] Decide how to integrate the tracy code into the scummvm build
* [ ] Integrate the `traceFrameEnd()` call into each backend to the "proper" place
* [ ] Mark threads with their names
* [ ] Implement MutexManger tracy integration
* [ ] OpenGL frame markers

### CODE REVIEW

File: configure [hunk 1]
```diff
@@ -5701,6 +5705,17 @@ echo "$_discord"
...
+#
+# Check whether to enable tracy profiler support
+#
+
+if test "$_tracy" = yes ; then
+	append_var DEFINES "-DTRACY_ENABLE"
+fi
+define_in_config_if_yes $_tracy 'ENABLE_TRACY'
+define_in_config_if_yes $_tracy 'TRACY_ON_DEMAND'
+define_in_config_if_yes $_tracy 'FORBIDDEN_SYMBOL_ALLOW_ALL'
```
Comment (mgerhardy) [inline configure:5717]: Hack until I know how to solve this in a better way.

File: devtools/create_project/create_project.cpp [hunk 2]
```diff
@@ -1070,6 +1070,7 @@ const Feature s_features[] = {
 	{      "translation",               "USE_TRANSLATION", false, true,  "Translation support" },
 	{           "vkeybd",                 "ENABLE_VKEYBD", false, false, "Virtual keyboard support"},
 	{    "eventrecorder",          "ENABLE_EVENTRECORDER", false, false, "Event recorder support"},
+	{            "tracy",                  "ENABLE_TRACY", false, false, "Tracy profiler support"},
```
Comment (mgerhardy) [inline devtools/create_project/create_project.cpp:1073]: This is also not yet done - will be finished if we can agree to include this - but `TRACY_ENABLE` must be given as define. That is not yet done here.

**Review by sev- — CHANGES_REQUESTED** [review]
I think this could be a good addition. I have a couple of question and notes, though.

File: .gitmodules [hunk 3]
```diff
@@ -2,3 +2,6 @@
 	path = devtools/create_prince
 	url = https://github.com/scummvm/game-translations
 	branch = prince-and-the-coward
+[submodule "common/tracy"]
```
[the remark this answers is not in the document]
Comment (mgerhardy) [reply .gitmodules:5]: Absolutely true - this is just for letting people play with it to decide whether a proper integration is worth the work. Just like to get feedback if I should continue to integrate it.

File: engines/twine/animations.cpp [hunk 4]
```diff
@@ -182,6 +183,7 @@ int32 Animations::getAnimMode(uint8 **ptr, const uint8 **keyFramePtr, const uint
 }
 
 bool Animations::setModelAnimation(int32 animState, const uint8 *animPtr, uint8 *bodyPtr, AnimTimerDataStruct *animTimerDataPtr) {
+	trace_scoped(setModelAnimation);
```
[the remark this answers is not in the document]
Comment (mgerhardy) [reply engines/twine/animations.cpp]: Added `traceScopedFunc` as the name must not always be the function name

File: common/trace.h [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+#ifndef COMMON_TRACE_H
+#define COMMON_TRACE_H
+
+#ifdef TRACY_ENABLE
+#include "tracy/Tracy.hpp"
+#endif
+
+#ifdef TRACY_ENABLE
+#define trace_value_scoped(name, x) ZoneNamedN(__tracy_scoped_##name, #name, true); ZoneValueV(__tracy_scoped_##name, (uint64_t)(x))
```
[the remark this answers is not in the document]
Comment (mgerhardy) [reply common/trace.h]: All renamed

File: common/trace.h [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#ifndef COMMON_TRACE_H
+#define COMMON_TRACE_H
+
+#ifdef TRACY_ENABLE
+#include "tracy/Tracy.hpp"
+#endif
```
[the remark this answers is not in the document]
Comment (mgerhardy) [reply common/trace.h]: Done.

File: .gitmodules (hunk 3 above)
[the remark this answers is not in the document]
Comment (mgerhardy) [reply .gitmodules:5]: The "problem" that I see here is that it's not a single c++ source file - but a lot of different headers and a lot of c++ files. Also the capture files are save, but the network protocol might differ from version to version. So it's the best to keep the profiler ui and the profiler "server" in sync somehow.

File: configure (hunk 1 above)
Comment (mgerhardy) [reply configure:5718]: Do you have any hint for me on how to solve this? I think that profiling should never get delivered in a release build anyway. So maybe activating this isn't a real problem for local builds?

On could just get rid of `common/trace.h` and do the

```c++
#ifdef TRACE_ENABLE
#include "common/tracy/Tracy.hpp"
#endif
```

everywhere in the c++ files where tracing is desired. But that would imo pollute the code with too much Tracy specific stuff. I'm really open for suggestions on how to solve this in a clean and nice way.

File: .gitmodules (hunk 3 above)
[the remark this answers is not in the document]
Comment (mgerhardy) [reply .gitmodules:5]: The author is suggesting this as integration method, too - https://github.com/wolfpld/tracy/blob/master/manual/tracy.tex#L388

One could also just assume that people who want to profile can copy the source files manually to the proper location and the build system will pick it up if that files exist. This wouldn't be a problem for make - but for msvc you would most likely have to re-run create_project for it.

File: configure (hunk 1 above)
Comment (mgerhardy) [reply configure:5718]: I just can't find a better way to do this. Is it a problem to activate this for those builds where tracy is enabled? I mean, tracy wouldn't be enabled for the CI builds, so errors would directly pop up (as `FORBIDDEN_SYMBOL_ALLOW_ALL` wouldn't be defined here of course.

**Comment by mgerhardy** [comment]
Closing this one - the profiler doesn't play very well with "inner-frames" - e.g. TwinE has a lot of do-while loops inside the main engine loop. Tracy needs a clear start and end of a frame - this doesn't work very nice with those inner-frames.

**Pull request closed without merging**